sdk/go/arvadosclient/arvadosclient.go

```diff
@@ -20,13 +20,31 @@ import (
 var MissingArvadosApiHost = errors.New("Missing required environment variable ARVADOS_API_HOST")
 var MissingArvadosApiToken = errors.New("Missing required environment variable ARVADOS_API_TOKEN")
 
-type ArvadosApiError struct {
-	error
-	HttpStatusCode int
-	HttpStatus string
+// Indicates an error that was returned by the API server.
+type RemoteApiServerError struct {
+	// Address of server returning error, of the form "host:port".
+	ServerAddress string
+
+	// Components of server response.
+	HttpStatusCode    int
+	HttpStatusMessage string
+
+	// Additional error details from response body.
+	ErrorDetails []string
 }
 
-func (e ArvadosApiError) Error() string { return e.error.Error() }
+func (e RemoteApiServerError) Error() string {
+	s := fmt.Sprintf("API server (%s) returned %d: %s.",
```

Comment:
Having all the information is certainly good but I feel like this formatting obscures the actual problem/error by putting it after a bunch of details.

    API server (zzzzz.arvadosapi.com:443) returned 422: Unprocessable Entity. Additional details: #<NoMemoryError: failed to allocate memory>

If we have ErrorDetails, the server has gone to the trouble of giving us a human-readable error string, so I think we should feature this first, and append the technical details -- starting with the more likely clues -- in case that message isn't enough.

    API error: #<NoMemoryError: failed to allocate memory> (422 Unprocessable Entity from zzzzz.arvadosapi.com:443)

Or, if we didn't get an error message in JSON:

    API error: 500 Internal Server Error from zzzzz.arvadosapi.com:443

Suggest dropping trailing period, cf. https://github.com/golang/go/wiki/CodeReviewComments#error-strings